dev/bots/custom_rules/analyze.dart

```diff
@@ -0,0 +1,78 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' show Directory;
+
+import 'package:analyzer/dart/analysis/analysis_context.dart';
+import 'package:analyzer/dart/analysis/analysis_context_collection.dart';
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/analysis/session.dart';
+import 'package:path/path.dart' as path;
+
+import '../utils.dart';
+
+/// Analyzes the given `flutterRootDirectory` with the given [AnalyzeRule]s.
+///
+/// If a compilation unit can not be resolved, this function ignores the
+/// corresponding dart source file and logs an error using [foundError].
+Future<void> analyzeDirectoryWithRules(String flutterRootDirectory, List<AnalyzeRule> rules) async {
```

Comment:
This seems a bit misleading as it cannot analyze an arbitrary directory. It can only analyze a directory that contains the Flutter framework in `package/flutter/lib`. Any reason this couldn't just apply the `rules` generically to the directory passed in (and `analyze.dart` would just pass in the lib directory of flutter framework)? Or be explicit in the naming (e.g. `analyzeFrameworkWithRules`) that it is _not_ able to analyze any directory?

Reply:
Sounds good. Maybe we can figure out how to generalize this once we have a use case for generalizing it :)